src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -385,25 +389,31 @@ internal static SecurityStatusPalErrorCode DoSslHandshake(SafeSslHandle context,
                 }
             }
 
-            int retVal = Ssl.SslDoHandshake(context);
-            if (retVal != 1)
+            try
             {
-                Exception? innerError;
-                Ssl.SslErrorCode error = GetSslError(context, retVal, out innerError);
-
-                if (error == Ssl.SslErrorCode.SSL_ERROR_WANT_X509_LOOKUP)
+                int retVal = Ssl.SslDoHandshake(context);
+                if (retVal != 1)
                 {
-                    return SecurityStatusPalErrorCode.CredentialsNeeded;
-                }
+                    Exception? innerError = null;
+                    Ssl.SslErrorCode error = GetSslError(context, retVal, out innerError);
 
-                if ((retVal != -1) || (error != Ssl.SslErrorCode.SSL_ERROR_WANT_READ))
-                {
-                    // Handshake failed, but even if the handshake does not need to read, there may be an Alert going out.
-                    // To handle that we will fall-through the block below to pull it out, and we will fail after.
-                    handshakeException = new SslException(SR.Format(SR.net_ssl_handshake_failed_error, error), innerError);
-                    Crypto.ErrClearError();
+                    if (error == Ssl.SslErrorCode.SSL_ERROR_WANT_X509_LOOKUP)
+                    {
+                        return SecurityStatusPalErrorCode.CredentialsNeeded;
+                    }
+
+                    if ((retVal != -1) || (error != Ssl.SslErrorCode.SSL_ERROR_WANT_READ))
+                    {
+                        // Handshake failed, but even if the handshake does not need to read, there may be an Alert going out.
+                        // To handle that we will fall-through the block below to pull it out, and we will fail after.
+                        handshakeException = new SslException(SR.Format(SR.net_ssl_handshake_failed_error, error), innerError);
+                    }
                 }
             }
+            finally
+            {
+                Crypto.ErrClearError();
```

Comment:
A comment would be helpful.

And there's no finer-grained place to do this?  Just seems a little bandaid-y to do this here, as it seems like we don't know where the extraneous error is coming from.

But if this is the best place, so be it.

Reply:
no, we know exactly what is failing - at least in this particular case.

```
139940077336320:error:1409442E:SSL routines:ssl3_read_bytes:tlsv1 alert protocol version:ssl/record/rec_layer_s3.c:1543:SSL alert number 70
```

The test is constructed to fail and sends TLS alert. But the client side is disposed before it is able to fully process it. 
It _should_ be sufficient to cleanup only on error or exception. 

The other aspect is that GetSslError() gets last error from OpenSSL. I don't tank if it is ever possible to get multiple errors. From that prospective `Crypto.ErrClearError` is generic bandaid for something that _may_ (or may not) happen. 

We _could_ avoid this by grabbing reference on the SafeHandle. If we do, we would do handshake AND have ability to query error and follow the error path. 

One more thought, Since the `Ssl.SslDoHandshake` is only called from here, we could change it to pass back also what ever GetSslError() would do. e.g. avoiding the second call that may throw.
```c#
Int Ssl.SslDoHandshake(SafeHandle context, ref Ssl.SslErrorCode error)
```

Reply:
Changing the PAL method to  combine two P/Invokes into one by making the error code an out/ref/pointer parameter, if they're always called in pairs, seems like some goodness (it'll reduce the interlocked counting on the safehandle, and all the other work that goes on in a P/Invoke).

As for sprinkling in a call to ERR_clear_error(): I'm actively working on a change to reverse all of these flows.  Every shim function that has a possibility of changing the error queue starts off with ERR_clear_error().  If there are multiple places that it might've sprinkled in noise, it'll call it multiple times.

Once that's done, it should be the case that we can remove the shim function for it altogether, because every flow looks like

* Clean queue
* Do work.
* Execption? Throw first error from the queue.
* No exception? Queue floats.
* Next op starts
* Clear error queue.
* Do work.
* ...

Reply:
> e.g. just calling GetError() to get last one may not be sufficient...

The new model is to throw the first one (clearing out the rest).  The reason the current/"old" model is the last one is because we ignored the queue and let it float around, so the last one is the only one we knew was relevant.  But, we've since learned of operations that emit multiple errors, the latter ones being bad error state recovery from the initial.

Reply:
I modified `SslDoHandshake` to give error back to avoid extra call and removed the try/finally.
That should be sufficient to deal with the released handle during handshake.